File: tests/src/tests/test_collision_filter.cpp
#include <catch2/catch_test_macros.hpp>

#include <ipc/collision_filter.hpp>

#include <Eigen/Core>
#include <vector>

using namespace ipc;

// ─────────────────────────────────────────────────────────────────────────────

TEST_CASE("CollisionFilter default", "[collision_filter]")
{
    CollisionFilter f;
    CHECK(f(0, 1));
    CHECK(f(5, 10));
    CHECK(f(0, 0)); // self-pair is accepted by default
}

// ─────────────────────────────────────────────────────────────────────────────

TEST_CASE("CollisionFilter from lambda", "[collision_filter]")
{
    // Two bodies: verts 0-2 in body 0, verts 3-5 in body 1.
    const std::vector<int> body = { 0, 0, 0, 1, 1, 1 };
    CollisionFilter self_col(
        [&body](size_t vi, size_t vj) { return body[vi] != body[vj]; });

    CHECK_FALSE(self_col(0, 1)); // same body
    CHECK_FALSE(self_col(3, 5)); // same body
    CHECK(self_col(0, 3));       // different bodies
    CHECK(self_col(2, 4));       // different bodies
}

// ─────────────────────────────────────────────────────────────────────────────

TEST_CASE("CollisionFilter operator!", "[collision_filter]")
{
    CollisionFilter same_body(
        [](size_t vi, size_t vj) { return vi / 3 == vj / 3; });

    CollisionFilter cross_body = !same_body;

    CHECK(same_body(0, 2));
    CHECK_FALSE(cross_body(0, 2));

    CHECK_FALSE(same_body(0, 4));
    CHECK(cross_body(0, 4));
}

// ─────────────────────────────────────────────────────────────────────────────

TEST_CASE("CollisionFilter operator|", "[collision_filter]")
{
    // Layer tags: 0,0,1,0,1,1
    const std::vector<int> layer = { 0, 0, 1, 0, 1, 1 };

    CollisionFilter a( // pass if vi is in layer 1
        [&layer](size_t vi, size_t /*vj*/) { return layer[vi] == 1; });
    CollisionFilter b( // pass if vj is in layer 0
        [&layer](size_t /*vi*/, size_t vj) { return layer[vj] == 0; });

    CollisionFilter a_or_b = a | b;

    CHECK(a_or_b(2, 5));       // a passes (vi=2, layer 1)
    CHECK(a_or_b(0, 3));       // b passes (vj=3, layer 0)
    CHECK_FALSE(a_or_b(0, 4)); // neither: vi=0 layer 0, vj=4 layer 1
}

// ─────────────────────────────────────────────────────────────────────────────

TEST_CASE("CollisionFilter operator&", "[collision_filter]")
{
    const std::vector<int> body = { 0, 0, 0, 1, 1, 1 };
    const std::vector<int> layer = { 0, 0, 1, 0, 1, 1 };

    CollisionFilter self_col(
        [&body](size_t vi, size_t vj) { return body[vi] != body[vj]; });
    CollisionFilter layer_col(
        [&layer](size_t vi, size_t vj) { return layer[vi] != layer[vj]; });

    CollisionFilter both = self_col & layer_col;

    // (0,1): same body → self_col blocks → intersection blocks
    CHECK_FALSE(both(0, 1));
    // (0,2): same body → self_col blocks → intersection blocks
    CHECK_FALSE(both(0, 2));
    // (0,3): cross body, same layer (0,0) → layer_col blocks → blocks
    CHECK_FALSE(both(0, 3));
    // (0,4): cross body, cross layer (0,1) → both pass
    CHECK(both(0, 4));
}

// ─────────────────────────────────────────────────────────────────────────────

TEST_CASE("CollisionFilter union/intersection semantics", "[collision_filter]")
{
    const std::vector<int> body = { 0, 0, 0, 1, 1, 1 };
    const std::vector<int> layer = { 0, 0, 1, 0, 1, 1 };

    CollisionFilter self_col(
        [&body](size_t vi, size_t vj) { return body[vi] != body[vj]; });
    CollisionFilter layer_col(
        [&layer](size_t vi, size_t vj) { return layer[vi] != layer[vj]; });

    CollisionFilter cf_union = self_col | layer_col;
    CollisionFilter cf_inter = self_col & layer_col;

    // (0,1): both block → union also blocks
    CHECK_FALSE(cf_union(0, 1));
    CHECK_FALSE(cf_inter(0, 1));

    // (0,2): self blocks, layer passes → union passes, inter blocks
    CHECK(cf_union(0, 2));
    CHECK_FALSE(cf_inter(0, 2));

    // (0,3): self passes, layer blocks → union passes, inter blocks
    CHECK(cf_union(0, 3));
    CHECK_FALSE(cf_inter(0, 3));

    // (0,4): both pass
    CHECK(cf_union(0, 4));
    CHECK(cf_inter(0, 4));
}

// ─────────────────────────────────────────────────────────────────────────────

TEST_CASE("CollisionFilter operator|=", "[collision_filter]")
{
    CollisionFilter f([](size_t vi, size_t /*vj*/) { return vi < 3; });
    f |= CollisionFilter([](size_t /*vi*/, size_t vj) { return vj < 3; });

    CHECK(f(5, 2));       // vj < 3
    CHECK(f(2, 5));       // vi < 3
    CHECK_FALSE(f(5, 5)); // neither
}

TEST_CASE("CollisionFilter operator&=", "[collision_filter]")
{
    CollisionFilter f([](size_t vi, size_t /*vj*/) { return vi < 10; });
    f &= CollisionFilter([](size_t /*vi*/, size_t vj) { return vj < 10; });

    CHECK(f(3, 4));        // both
    CHECK_FALSE(f(3, 15)); // vj >= 10
    CHECK_FALSE(f(15, 3)); // vi >= 10
}

// ─────────────────────────────────────────────────────────────────────────────

TEST_CASE("make_vertex_patches_filter", "[collision_filter]")
{
    Eigen::VectorXi patches(6);
    patches << 0, 0, 0, 1, 1, 1;

    CollisionFilter f = make_vertex_patches_filter(patches);

    CHECK(f(0, 3));       // different patches
    CHECK(f(2, 4));       // different patches
    CHECK_FALSE(f(0, 1)); // same patch 0
    CHECK_FALSE(f(3, 5)); // same patch 1
}

// ─────────────────────────────────────────────────────────────────────────────

TEST_CASE("make_static_obstacle_filter", "[collision_filter]")
{
    // Vertices 0-3 are dynamic; vertices 4-5 are static obstacles.
    const size_t n_dynamic = 4;
    CollisionFilter f = make_static_obstacle_filter(n_dynamic);

    // dynamic–dynamic: allowed
    CHECK(f(0, 3));
    CHECK(f(1, 2));

    // dynamic–static: allowed
    CHECK(f(2, 4));
    CHECK(f(0, 5));
    CHECK(f(4, 1)); // order flipped

    // static–static: blocked
    CHECK_FALSE(f(4, 5));
    CHECK_FALSE(f(5, 4));
}

// ─────────────────────────────────────────────────────────────────────────────

TEST_CASE("make_connected_components_filter", "[collision_filter]")
{
    // Two disconnected triangles: {0,1,2} and {3,4,5}
    Eigen::MatrixXi faces(2, 3);
    faces << 0, 1, 2, 3, 4, 5;

    CollisionFilter f = make_connected_components_filter(faces);

    // Cross-component pairs: allowed
    CHECK(f(0, 3));
    CHECK(f(1, 4));
    CHECK(f(2, 5));

    // Intra-component pairs: blocked
    CHECK_FALSE(f(0, 1));
    CHECK_FALSE(f(0, 2));
    CHECK_FALSE(f(1, 2));
    CHECK_FALSE(f(3, 4));
    CHECK_FALSE(f(3, 5));
    CHECK_FALSE(f(4, 5));
}

// ─────────────────────────────────────────────────────────────────────────────

TEST_CASE("CollisionFilter composition chain", "[collision_filter]")
{
    // Combine: cross-component AND not-static-static
    Eigen::MatrixXi faces(2, 3);
    faces << 0, 1, 2, 3, 4, 5;

    // Verts 4 and 5 are also "static obstacles"
    const size_t n_dynamic = 4;

    CollisionFilter no_self = make_connected_components_filter(faces);
    CollisionFilter no_static = make_static_obstacle_filter(n_dynamic);

    CollisionFilter active = no_self & no_static;

    // Cross-component, at least one dynamic → allowed
    CHECK(active(0, 3));
    CHECK(active(1, 4)); // vj=4 is static but vi=1 is dynamic
    CHECK(active(2, 5)); // vj=5 is static but vi=2 is dynamic

    // Static–static cross-component pair → blocked by no_static
    CHECK_FALSE(active(4, 5));

    // Intra-component → blocked by no_self
    CHECK_FALSE(active(0, 1));
    CHECK_FALSE(active(3, 5));
}

File: python/tests/test_collision_filter.py
import find_ipctk
import numpy as np
from ipctk import (
    CollisionFilter,
    CollisionMesh,
    make_connected_components_filter,
    make_sparse_filter,
    make_static_obstacle_filter,
    make_vertex_patches_filter,
)

# ─────────────────────────────────────────────────────────────────────────────
# CollisionFilter basic construction
# ─────────────────────────────────────────────────────────────────────────────


def test_collision_filter_default():
    """Default-constructed filter accepts all pairs."""
    f = CollisionFilter()
    assert f(0, 1)
    assert f(5, 10)
    assert f(0, 0)  # self-pair accepted by default


def test_collision_filter_from_callable():
    """Filter constructed from a Python callable."""
    # Two bodies: verts 0-2 in body 0, verts 3-5 in body 1.
    body = [0, 0, 0, 1, 1, 1]
    f = CollisionFilter(lambda i, j: body[i] != body[j])

    assert not f(0, 1)  # same body
    assert not f(3, 5)  # same body
    assert f(0, 3)  # different bodies
    assert f(2, 4)  # different bodies


# ─────────────────────────────────────────────────────────────────────────────
# Composition operators
# ─────────────────────────────────────────────────────────────────────────────


def test_collision_filter_negation():
    """~ inverts the filter result."""
    body = [0, 0, 0, 1, 1, 1]
    same_body = CollisionFilter(lambda i, j: body[i] == body[j])
    cross_body = ~same_body

    assert same_body(0, 2)
    assert not cross_body(0, 2)

    assert not same_body(0, 4)
    assert cross_body(0, 4)


def test_collision_filter_union():
    """f | g accepts a pair if EITHER filter passes."""
    # Layer tags: 0,0,1,0,1,1
    layer = [0, 0, 1, 0, 1, 1]

    a = CollisionFilter(lambda i, j: layer[i] == 1)  # pass if vi in layer 1
    b = CollisionFilter(lambda i, j: layer[j] == 0)  # pass if vj in layer 0

    a_or_b = a | b

    assert a_or_b(2, 5)  # a passes (vi=2, layer 1)
    assert a_or_b(0, 3)  # b passes (vj=3, layer 0)
    assert not a_or_b(0, 4)  # neither: vi=0 layer 0, vj=4 layer 1


def test_collision_filter_intersection():
    """f & g accepts a pair only if BOTH filters pass."""
    body = [0, 0, 0, 1, 1, 1]
    layer = [0, 0, 1, 0, 1, 1]

    self_col = CollisionFilter(lambda i, j: body[i] != body[j])
    layer_col = CollisionFilter(lambda i, j: layer[i] != layer[j])

    both = self_col & layer_col

    # (0,1): same body → self_col blocks → intersection blocks
    assert not both(0, 1)
    # (0,2): same body → self_col blocks → intersection blocks
    assert not both(0, 2)
    # (0,3): cross body, same layer (0,0) → layer_col blocks → blocks
    assert not both(0, 3)
    # (0,4): cross body, cross layer (0,1) → both pass
    assert both(0, 4)


def test_collision_filter_union_intersection_semantics():
    """Union passes when either passes; intersection requires both."""
    body = [0, 0, 0, 1, 1, 1]
    layer = [0, 0, 1, 0, 1, 1]

    self_col = CollisionFilter(lambda i, j: body[i] != body[j])
    layer_col = CollisionFilter(lambda i, j: layer[i] != layer[j])

    cf_union = self_col | layer_col
    cf_inter = self_col & layer_col

    # (0,1): both block → union and intersection both block
    assert not cf_union(0, 1)
    assert not cf_inter(0, 1)

    # (0,2): self blocks, layer passes → union passes, inter blocks
    assert cf_union(0, 2)
    assert not cf_inter(0, 2)

    # (0,3): self passes, layer blocks → union passes, inter blocks
    assert cf_union(0, 3)
    assert not cf_inter(0, 3)

    # (0,4): both pass
    assert cf_union(0, 4)
    assert cf_inter(0, 4)


def test_collision_filter_ior():
    """f |= g modifies f to be the union of f and g."""
    f = CollisionFilter(lambda i, j: i < 3)
    f |= CollisionFilter(lambda i, j: j < 3)

    assert f(5, 2)  # j < 3
    assert f(2, 5)  # i < 3
    assert not f(5, 5)  # neither


def test_collision_filter_iand():
    """f &= g modifies f to be the intersection of f and g."""
    f = CollisionFilter(lambda i, j: i < 10)
    f &= CollisionFilter(lambda i, j: j < 10)

    assert f(3, 4)  # both
    assert not f(3, 15)  # j >= 10
    assert not f(15, 3)  # i >= 10


# ─────────────────────────────────────────────────────────────────────────────
# Factory functions
# ─────────────────────────────────────────────────────────────────────────────


def test_make_vertex_patches_filter():
    """Blocks pairs within the same patch, allows cross-patch pairs."""
    patches = np.array([0, 0, 0, 1, 1, 1], dtype=np.int32)
    f = make_vertex_patches_filter(patches)

    assert f(0, 3)  # different patches
    assert f(2, 4)  # different patches
    assert not f(0, 1)  # same patch 0
    assert not f(3, 5)  # same patch 1


def test_make_static_obstacle_filter():
    """Blocks static-static pairs; allows dynamic-* pairs."""
    # Vertices 0-3 are dynamic; vertices 4-5 are static obstacles.
    n_dynamic = 4
    f = make_static_obstacle_filter(n_dynamic)

    # dynamic–dynamic: allowed
    assert f(0, 3)
    assert f(1, 2)

    # dynamic–static: allowed (either order)
    assert f(2, 4)
    assert f(0, 5)
    assert f(4, 1)  # order flipped

    # static–static: blocked
    assert not f(4, 5)
    assert not f(5, 4)


def test_make_connected_components_filter():
    """Blocks intra-component pairs; allows cross-component pairs."""
    # Two disconnected triangles: {0,1,2} and {3,4,5}
    faces = np.array([[0, 1, 2], [3, 4, 5]], dtype=np.int32)
    f = make_connected_components_filter(faces)

    # Cross-component: allowed
    assert f(0, 3)
    assert f(1, 4)
    assert f(2, 5)

    # Intra-component: blocked
    assert not f(0, 1)
    assert not f(0, 2)
    assert not f(1, 2)
    assert not f(3, 4)
    assert not f(3, 5)
    assert not f(4, 5)


# ─────────────────────────────────────────────────────────────────────────────
# Composition chain
# ─────────────────────────────────────────────────────────────────────────────


def test_collision_filter_composition_chain():
    """Combine connected-components and static-obstacle filters."""
    # Two disconnected triangles: {0,1,2} and {3,4,5}
    faces = np.array([[0, 1, 2], [3, 4, 5]], dtype=np.int32)

    # Vertices 4 and 5 are also "static obstacles"
    n_dynamic = 4

    no_self = make_connected_components_filter(faces)
    no_static = make_static_obstacle_filter(n_dynamic)

    active = no_self & no_static

    # Cross-component, at least one dynamic → allowed
    assert active(0, 3)
    assert active(1, 4)  # vj=4 is static but vi=1 is dynamic
    assert active(2, 5)  # vj=5 is static but vi=2 is dynamic

    # Static–static cross-component pair → blocked by no_static
    assert not active(4, 5)

    # Intra-component → blocked by no_self
    assert not active(0, 1)
    assert not active(3, 5)


# ─────────────────────────────────────────────────────────────────────────────
# SparseCanCollide composition
# ─────────────────────────────────────────────────────────────────────────────


def test_make_sparse_filter():
    """make_sparse_filter returns a composable CollisionFilter."""
    # Block pair (0, 1) explicitly; allow everything else.
    sparse = make_sparse_filter({(0, 1): False}, True)

    assert sparse(0, 2)      # not in map → default True
    assert not sparse(0, 1)  # explicit False

    # ~ (invert): (0,1) now passes; everything else is blocked
    inverted = ~sparse
    assert inverted(0, 1)      # was blocked, now passes
    assert not inverted(0, 2)  # was allowed, now blocked

    # | (union): the blocked (0,1) pair is rescued by a second filter
    rescue_01 = CollisionFilter(lambda i, j: i == 0 and j == 1)
    rescued = sparse | rescue_01
    assert rescued(0, 1)  # sparse blocks but rescue_01 passes → union passes
    assert rescued(0, 2)  # sparse passes → union passes

    # & (intersection): add a restriction on top of sparse
    only_small = CollisionFilter(lambda i, j: i < 3 and j < 3)
    restricted = sparse & only_small
    assert restricted(0, 2)      # sparse allows, only_small allows → passes
    assert not restricted(0, 1)  # sparse blocks → intersection fails
    assert not restricted(0, 5)  # only_small fails (j=5 ≥ 3) → intersection fails


def test_collision_filter_on_mesh_can_collide():
    """CollisionFilter assigned to mesh.can_collide works correctly."""
    V = np.array(
        [[0, 0], [1, 0], [0, 1], [1, 1], [2, 0], [3, 0], [2, 1], [3, 1]], dtype=float
    )
    E = np.array(
        [[0, 1], [1, 3], [3, 2], [2, 0], [4, 5], [5, 7], [7, 6], [6, 4]], dtype=int
    )
    mesh = CollisionMesh(V, E)

    patches = np.array([0, 0, 0, 0, 1, 1, 1, 1], dtype=np.int32)
    f = make_vertex_patches_filter(patches)

    mesh.can_collide = f

    for i in range(V.shape[0]):
        for j in range(V.shape[0]):
            assert mesh.can_collide(i, j) == (patches[i] != patches[j])
